// @rocketh/viem/dist/index.js
import { logs } from 'named-logs';
import { createPublicClient, createWalletClient, custom, getContract as getViemContract, } from 'viem';
const logger = logs('@rocketh/viem');
export function viem(env) {
    const walletClient = createWalletClient({
        chain: env.network.chain,
        transport: custom(env.network.provider),
    });
    const publicClient = createPublicClient({
        chain: env.network.chain,
        transport: custom(env.network.provider),
    });
    return {
        walletClient,
        publicClient,
        getWritableContract(name) {
            const deployment = typeof name === 'string' ? env.get(name) : name;
            return getViemContract({
                address: deployment.address,
                abi: deployment.abi,
                client: { public: publicClient, wallet: walletClient },
            });
        },
        getContract(name) {
            const deployment = typeof name === 'string' ? env.get(name) : name;
            return getViemContract({
                address: deployment.address,
                abi: deployment.abi,
                client: { public: publicClient },
            });
        },
    };
}
