// @ensdomains/hardhat-chai-matchers-viem/dist/matchers/utils/getReturnDataFromError.js
import { decodeErrorResult, } from "viem";
import { getKnownPanicReason } from "../../constants.js";
import { formatArgs } from "./formatArgs.js";
export const isWalkableError = (err) => typeof err === "object" &&
    err !== null &&
    "walk" in err &&
    typeof err["walk"] === "function";
export const getRawErrorData = (error) => {
    if (typeof error.data !== "object")
        return error.data;
    if (typeof error.data.data === "string")
        return error.data.data;
    if ("errorName" in error.data &&
        "args" in error.data &&
        "abiItem" in error.data)
        return error.data;
    return undefined;
};
const tryDecodeReturnData = ({ abi, data }) => {
    try {
        return decodeErrorResult({ abi, data });
    }
    catch {
        return null;
    }
};
export const errorWhy = {
    empty: "transaction reverted without a reason",
    unknownContract: "transaction reverted with unknown error",
    panic: (code, description) => `transaction reverted with panic code: ${code} (${description})`,
    error: (reason) => `transaction reverted with string: ${reason}`,
    custom: (name, args) => `transaction reverted with error: ${name}(${formatArgs(args)})`,
};
export const getReturnDataFromError = (subject, error) => {
    if (!isWalkableError(error))
        return { kind: "unknown-local" };
    const sourceError = error.walk();
    const raw = getRawErrorData(sourceError);
    if (!raw)
        return { kind: "unknown-local" };
    if (raw === "0x")
        return { kind: "empty", sourceError, why: errorWhy.empty };
    const decodedReturnData = typeof raw === "object"
        ? raw
        : tryDecodeReturnData({
            abi: subject.abi,
            data: raw,
        });
    if (!decodedReturnData)
        return {
            kind: "unknown-contract",
            sourceError,
            why: errorWhy.unknownContract,
        };
    if (decodedReturnData.errorName === "Panic") {
        const code = decodedReturnData.args[0];
        return {
            kind: "panic",
            code,
            description: getKnownPanicReason(code),
            sourceError,
            why: errorWhy.panic(code, getKnownPanicReason(code)),
        };
    }
    if (decodedReturnData.errorName === "Error") {
        return {
            kind: "error",
            reason: decodedReturnData.args[0],
            sourceError,
            why: errorWhy.error(decodedReturnData.args[0]),
        };
    }
    return {
        kind: "custom",
        name: decodedReturnData.errorName,
        args: decodedReturnData.args,
        sourceError,
        why: errorWhy.custom(decodedReturnData.errorName, decodedReturnData.args),
    };
};
