// @ensdomains/hardhat-chai-matchers-viem/dist/internal/hook-handlers/network.js
import { sendTransaction } from "viem/actions";
import { deployCustomContract } from "../../utils/deployCustomContract.js";
import { addChaiMatchers } from "../addChaiMatchers.js";
let isInitialized = false;
// @ts-expect-error - __vitest_worker__ is not typed
const isVitestEnvironment = () => globalThis["__vitest_worker__"] !== undefined;
async function networkHookHandler() {
    const handlers = {
        async newConnection(context, next) {
            if (!isInitialized && isVitestEnvironment()) {
                await addChaiMatchers();
                isInitialized = true;
            }
            const connection = await next(context);
            const publicClient = await connection.viem.getPublicClient();
            const originalDeployContract = connection.viem.deployContract;
            const originalGetContractAt = connection.viem.getContractAt;
            const createFunctionProxy = async (originalFunction, ...args) => {
                const client = (() => {
                    const lastArg = args[args.length - 1];
                    if (typeof lastArg === "object" &&
                        "client" in lastArg &&
                        "public" in lastArg.client)
                        return lastArg.client.public;
                    return publicClient;
                })();
                const result = await originalFunction(...args);
                const createMetadataProxy = (original, kind) => new Proxy({}, {
                    get(_, functionName) {
                        return (...args) => {
                            const resultPromise = original[functionName](...args);
                            resultPromise.__call_metadata = {
                                functionName,
                                args,
                                client,
                                kind,
                                abi: result.abi,
                                address: result.address,
                            };
                            return resultPromise;
                        };
                    },
                });
                if ("write" in result) {
                    const originalWrite = result.write;
                    result.write = createMetadataProxy(originalWrite, "write");
                }
                if ("read" in result) {
                    const originalRead = result.read;
                    result.read = createMetadataProxy(originalRead, "read");
                }
                result.arbitrary = (args) => {
                    const resultPromise = sendTransaction(publicClient, args);
                    resultPromise.__call_metadata = {
                        functionName: ".arbitrary",
                        args,
                        client: publicClient,
                        kind: "arbitrary",
                        abi: result.abi,
                        address: result.address,
                    };
                    return resultPromise;
                };
                result.client = publicClient;
                return result;
            };
            connection.viem.deployContract = async (...args) => {
                if (typeof args[0] === "string")
                    return createFunctionProxy(originalDeployContract, ...args);
                return createFunctionProxy(deployCustomContract, connection, ...args);
            };
            connection.viem.getContractAt = async (...args) => createFunctionProxy(originalGetContractAt, ...args);
            return connection;
        },
    };
    return handlers;
}
export default networkHookHandler;
