// @ensdomains/address-encoder/dist/cjs/coins.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.gxc = exports.grin = exports.goLegacy = exports.gnoLegacy = exports.ftmLegacy = exports.flux = exports.flow = exports.firo = exports.fio = exports.fil = exports.ewtLegacy = exports.etn = exports.eth = exports.etcLegacy = exports.eos = exports.ela = exports.egld = exports.dot = exports.doge = exports.divi = exports.dgb = exports.dcr = exports.dash = exports.cloLegacy = exports.ckb = exports.celoLegacy = exports.ccxx = exports.cca = exports.bts = exports.btm = exports.btg = exports.btc = exports.bsv = exports.bps = exports.bnb = exports.bdx = exports.bcn = exports.bch = exports.bcd = exports.avax = exports.atom = exports.ark = exports.ardr = exports.ar = exports.algo = exports.aion = exports.aib = exports.ae = exports.ada = exports.abbc = void 0;
exports.trx = exports.tomoLegacy = exports.thetaLegacy = exports.tfuel = exports.sys = exports.stx = exports.strk = exports.strat = exports.steem = exports.srm = exports.sol = exports.sero = exports.sc = exports.rvn = exports.rune = exports.rdd = exports.rbtc = exports.qtum = exports.ppc = exports.poaLegacy = exports.ont = exports.one = exports.nuls = exports.nrgLegacy = exports.nostr = exports.nmc = exports.nim = exports.neo = exports.near = exports.nas = exports.nano = exports.mrx = exports.mona = exports.luna = exports.ltc = exports.lsk = exports.lrg = exports.lcc = exports.ksm = exports.kmd = exports.kava = exports.iris = exports.iotx = exports.iota = exports.iost = exports.icx = exports.hnt = exports.hns = exports.hive = exports.hbar = void 0;
exports.zil = exports.zen = exports.zec = exports.xvg = exports.xtz = exports.xrp = exports.xmr = exports.xlm = exports.xhv = exports.xem = exports.xch = exports.wicc = exports.waves = exports.wan = exports.vsys = exports.vlxLegacy = exports.vlx = exports.via = exports.vet = exports.ttLegacy = void 0;
var abbc_js_1 = require("./coin/abbc.js");
Object.defineProperty(exports, "abbc", { enumerable: true, get: function () { return abbc_js_1.abbc; } });
var ada_js_1 = require("./coin/ada.js");
Object.defineProperty(exports, "ada", { enumerable: true, get: function () { return ada_js_1.ada; } });
var ae_js_1 = require("./coin/ae.js");
Object.defineProperty(exports, "ae", { enumerable: true, get: function () { return ae_js_1.ae; } });
var aib_js_1 = require("./coin/aib.js");
Object.defineProperty(exports, "aib", { enumerable: true, get: function () { return aib_js_1.aib; } });
var aion_js_1 = require("./coin/aion.js");
Object.defineProperty(exports, "aion", { enumerable: true, get: function () { return aion_js_1.aion; } });
var algo_js_1 = require("./coin/algo.js");
Object.defineProperty(exports, "algo", { enumerable: true, get: function () { return algo_js_1.algo; } });
var ar_js_1 = require("./coin/ar.js");
Object.defineProperty(exports, "ar", { enumerable: true, get: function () { return ar_js_1.ar; } });
var ardr_js_1 = require("./coin/ardr.js");
Object.defineProperty(exports, "ardr", { enumerable: true, get: function () { return ardr_js_1.ardr; } });
var ark_js_1 = require("./coin/ark.js");
Object.defineProperty(exports, "ark", { enumerable: true, get: function () { return ark_js_1.ark; } });
var atom_js_1 = require("./coin/atom.js");
Object.defineProperty(exports, "atom", { enumerable: true, get: function () { return atom_js_1.atom; } });
var avax_js_1 = require("./coin/avax.js");
Object.defineProperty(exports, "avax", { enumerable: true, get: function () { return avax_js_1.avax; } });
var bcd_js_1 = require("./coin/bcd.js");
Object.defineProperty(exports, "bcd", { enumerable: true, get: function () { return bcd_js_1.bcd; } });
var bch_js_1 = require("./coin/bch.js");
Object.defineProperty(exports, "bch", { enumerable: true, get: function () { return bch_js_1.bch; } });
var bcn_js_1 = require("./coin/bcn.js");
Object.defineProperty(exports, "bcn", { enumerable: true, get: function () { return bcn_js_1.bcn; } });
var bdx_js_1 = require("./coin/bdx.js");
Object.defineProperty(exports, "bdx", { enumerable: true, get: function () { return bdx_js_1.bdx; } });
var bnb_js_1 = require("./coin/bnb.js");
Object.defineProperty(exports, "bnb", { enumerable: true, get: function () { return bnb_js_1.bnb; } });
var bps_js_1 = require("./coin/bps.js");
Object.defineProperty(exports, "bps", { enumerable: true, get: function () { return bps_js_1.bps; } });
var bsv_js_1 = require("./coin/bsv.js");
Object.defineProperty(exports, "bsv", { enumerable: true, get: function () { return bsv_js_1.bsv; } });
var btc_js_1 = require("./coin/btc.js");
Object.defineProperty(exports, "btc", { enumerable: true, get: function () { return btc_js_1.btc; } });
var btg_js_1 = require("./coin/btg.js");
Object.defineProperty(exports, "btg", { enumerable: true, get: function () { return btg_js_1.btg; } });
var btm_js_1 = require("./coin/btm.js");
Object.defineProperty(exports, "btm", { enumerable: true, get: function () { return btm_js_1.btm; } });
var bts_js_1 = require("./coin/bts.js");
Object.defineProperty(exports, "bts", { enumerable: true, get: function () { return bts_js_1.bts; } });
var cca_js_1 = require("./coin/cca.js");
Object.defineProperty(exports, "cca", { enumerable: true, get: function () { return cca_js_1.cca; } });
var ccxx_js_1 = require("./coin/ccxx.js");
Object.defineProperty(exports, "ccxx", { enumerable: true, get: function () { return ccxx_js_1.ccxx; } });
var celoLegacy_js_1 = require("./coin/celoLegacy.js");
Object.defineProperty(exports, "celoLegacy", { enumerable: true, get: function () { return celoLegacy_js_1.celoLegacy; } });
var ckb_js_1 = require("./coin/ckb.js");
Object.defineProperty(exports, "ckb", { enumerable: true, get: function () { return ckb_js_1.ckb; } });
var cloLegacy_js_1 = require("./coin/cloLegacy.js");
Object.defineProperty(exports, "cloLegacy", { enumerable: true, get: function () { return cloLegacy_js_1.cloLegacy; } });
var dash_js_1 = require("./coin/dash.js");
Object.defineProperty(exports, "dash", { enumerable: true, get: function () { return dash_js_1.dash; } });
var dcr_js_1 = require("./coin/dcr.js");
Object.defineProperty(exports, "dcr", { enumerable: true, get: function () { return dcr_js_1.dcr; } });
var dgb_js_1 = require("./coin/dgb.js");
Object.defineProperty(exports, "dgb", { enumerable: true, get: function () { return dgb_js_1.dgb; } });
var divi_js_1 = require("./coin/divi.js");
Object.defineProperty(exports, "divi", { enumerable: true, get: function () { return divi_js_1.divi; } });
var doge_js_1 = require("./coin/doge.js");
Object.defineProperty(exports, "doge", { enumerable: true, get: function () { return doge_js_1.doge; } });
var dot_js_1 = require("./coin/dot.js");
Object.defineProperty(exports, "dot", { enumerable: true, get: function () { return dot_js_1.dot; } });
var egld_js_1 = require("./coin/egld.js");
Object.defineProperty(exports, "egld", { enumerable: true, get: function () { return egld_js_1.egld; } });
var ela_js_1 = require("./coin/ela.js");
Object.defineProperty(exports, "ela", { enumerable: true, get: function () { return ela_js_1.ela; } });
var eos_js_1 = require("./coin/eos.js");
Object.defineProperty(exports, "eos", { enumerable: true, get: function () { return eos_js_1.eos; } });
var etcLegacy_js_1 = require("./coin/etcLegacy.js");
Object.defineProperty(exports, "etcLegacy", { enumerable: true, get: function () { return etcLegacy_js_1.etcLegacy; } });
var eth_js_1 = require("./coin/eth.js");
Object.defineProperty(exports, "eth", { enumerable: true, get: function () { return eth_js_1.eth; } });
var etn_js_1 = require("./coin/etn.js");
Object.defineProperty(exports, "etn", { enumerable: true, get: function () { return etn_js_1.etn; } });
var ewtLegacy_js_1 = require("./coin/ewtLegacy.js");
Object.defineProperty(exports, "ewtLegacy", { enumerable: true, get: function () { return ewtLegacy_js_1.ewtLegacy; } });
var fil_js_1 = require("./coin/fil.js");
Object.defineProperty(exports, "fil", { enumerable: true, get: function () { return fil_js_1.fil; } });
var fio_js_1 = require("./coin/fio.js");
Object.defineProperty(exports, "fio", { enumerable: true, get: function () { return fio_js_1.fio; } });
var firo_js_1 = require("./coin/firo.js");
Object.defineProperty(exports, "firo", { enumerable: true, get: function () { return firo_js_1.firo; } });
var flow_js_1 = require("./coin/flow.js");
Object.defineProperty(exports, "flow", { enumerable: true, get: function () { return flow_js_1.flow; } });
var flux_js_1 = require("./coin/flux.js");
Object.defineProperty(exports, "flux", { enumerable: true, get: function () { return flux_js_1.flux; } });
var ftmLegacy_js_1 = require("./coin/ftmLegacy.js");
Object.defineProperty(exports, "ftmLegacy", { enumerable: true, get: function () { return ftmLegacy_js_1.ftmLegacy; } });
var gnoLegacy_js_1 = require("./coin/gnoLegacy.js");
Object.defineProperty(exports, "gnoLegacy", { enumerable: true, get: function () { return gnoLegacy_js_1.gnoLegacy; } });
var goLegacy_js_1 = require("./coin/goLegacy.js");
Object.defineProperty(exports, "goLegacy", { enumerable: true, get: function () { return goLegacy_js_1.goLegacy; } });
var grin_js_1 = require("./coin/grin.js");
Object.defineProperty(exports, "grin", { enumerable: true, get: function () { return grin_js_1.grin; } });
var gxc_js_1 = require("./coin/gxc.js");
Object.defineProperty(exports, "gxc", { enumerable: true, get: function () { return gxc_js_1.gxc; } });
var hbar_js_1 = require("./coin/hbar.js");
Object.defineProperty(exports, "hbar", { enumerable: true, get: function () { return hbar_js_1.hbar; } });
var hive_js_1 = require("./coin/hive.js");
Object.defineProperty(exports, "hive", { enumerable: true, get: function () { return hive_js_1.hive; } });
var hns_js_1 = require("./coin/hns.js");
Object.defineProperty(exports, "hns", { enumerable: true, get: function () { return hns_js_1.hns; } });
var hnt_js_1 = require("./coin/hnt.js");
Object.defineProperty(exports, "hnt", { enumerable: true, get: function () { return hnt_js_1.hnt; } });
var icx_js_1 = require("./coin/icx.js");
Object.defineProperty(exports, "icx", { enumerable: true, get: function () { return icx_js_1.icx; } });
var iost_js_1 = require("./coin/iost.js");
Object.defineProperty(exports, "iost", { enumerable: true, get: function () { return iost_js_1.iost; } });
var iota_js_1 = require("./coin/iota.js");
Object.defineProperty(exports, "iota", { enumerable: true, get: function () { return iota_js_1.iota; } });
var iotx_js_1 = require("./coin/iotx.js");
Object.defineProperty(exports, "iotx", { enumerable: true, get: function () { return iotx_js_1.iotx; } });
var iris_js_1 = require("./coin/iris.js");
Object.defineProperty(exports, "iris", { enumerable: true, get: function () { return iris_js_1.iris; } });
var kava_js_1 = require("./coin/kava.js");
Object.defineProperty(exports, "kava", { enumerable: true, get: function () { return kava_js_1.kava; } });
var kmd_js_1 = require("./coin/kmd.js");
Object.defineProperty(exports, "kmd", { enumerable: true, get: function () { return kmd_js_1.kmd; } });
var ksm_js_1 = require("./coin/ksm.js");
Object.defineProperty(exports, "ksm", { enumerable: true, get: function () { return ksm_js_1.ksm; } });
var lcc_js_1 = require("./coin/lcc.js");
Object.defineProperty(exports, "lcc", { enumerable: true, get: function () { return lcc_js_1.lcc; } });
var lrg_js_1 = require("./coin/lrg.js");
Object.defineProperty(exports, "lrg", { enumerable: true, get: function () { return lrg_js_1.lrg; } });
var lsk_js_1 = require("./coin/lsk.js");
Object.defineProperty(exports, "lsk", { enumerable: true, get: function () { return lsk_js_1.lsk; } });
var ltc_js_1 = require("./coin/ltc.js");
Object.defineProperty(exports, "ltc", { enumerable: true, get: function () { return ltc_js_1.ltc; } });
var luna_js_1 = require("./coin/luna.js");
Object.defineProperty(exports, "luna", { enumerable: true, get: function () { return luna_js_1.luna; } });
var mona_js_1 = require("./coin/mona.js");
Object.defineProperty(exports, "mona", { enumerable: true, get: function () { return mona_js_1.mona; } });
var mrx_js_1 = require("./coin/mrx.js");
Object.defineProperty(exports, "mrx", { enumerable: true, get: function () { return mrx_js_1.mrx; } });
var nano_js_1 = require("./coin/nano.js");
Object.defineProperty(exports, "nano", { enumerable: true, get: function () { return nano_js_1.nano; } });
var nas_js_1 = require("./coin/nas.js");
Object.defineProperty(exports, "nas", { enumerable: true, get: function () { return nas_js_1.nas; } });
var near_js_1 = require("./coin/near.js");
Object.defineProperty(exports, "near", { enumerable: true, get: function () { return near_js_1.near; } });
var neo_js_1 = require("./coin/neo.js");
Object.defineProperty(exports, "neo", { enumerable: true, get: function () { return neo_js_1.neo; } });
var nim_js_1 = require("./coin/nim.js");
Object.defineProperty(exports, "nim", { enumerable: true, get: function () { return nim_js_1.nim; } });
var nmc_js_1 = require("./coin/nmc.js");
Object.defineProperty(exports, "nmc", { enumerable: true, get: function () { return nmc_js_1.nmc; } });
var nostr_js_1 = require("./coin/nostr.js");
Object.defineProperty(exports, "nostr", { enumerable: true, get: function () { return nostr_js_1.nostr; } });
var nrgLegacy_js_1 = require("./coin/nrgLegacy.js");
Object.defineProperty(exports, "nrgLegacy", { enumerable: true, get: function () { return nrgLegacy_js_1.nrgLegacy; } });
var nuls_js_1 = require("./coin/nuls.js");
Object.defineProperty(exports, "nuls", { enumerable: true, get: function () { return nuls_js_1.nuls; } });
var one_js_1 = require("./coin/one.js");
Object.defineProperty(exports, "one", { enumerable: true, get: function () { return one_js_1.one; } });
var ont_js_1 = require("./coin/ont.js");
Object.defineProperty(exports, "ont", { enumerable: true, get: function () { return ont_js_1.ont; } });
var poaLegacy_js_1 = require("./coin/poaLegacy.js");
Object.defineProperty(exports, "poaLegacy", { enumerable: true, get: function () { return poaLegacy_js_1.poaLegacy; } });
var ppc_js_1 = require("./coin/ppc.js");
Object.defineProperty(exports, "ppc", { enumerable: true, get: function () { return ppc_js_1.ppc; } });
var qtum_js_1 = require("./coin/qtum.js");
Object.defineProperty(exports, "qtum", { enumerable: true, get: function () { return qtum_js_1.qtum; } });
var rbtc_js_1 = require("./coin/rbtc.js");
Object.defineProperty(exports, "rbtc", { enumerable: true, get: function () { return rbtc_js_1.rbtc; } });
var rdd_js_1 = require("./coin/rdd.js");
Object.defineProperty(exports, "rdd", { enumerable: true, get: function () { return rdd_js_1.rdd; } });
var rune_js_1 = require("./coin/rune.js");
Object.defineProperty(exports, "rune", { enumerable: true, get: function () { return rune_js_1.rune; } });
var rvn_js_1 = require("./coin/rvn.js");
Object.defineProperty(exports, "rvn", { enumerable: true, get: function () { return rvn_js_1.rvn; } });
var sc_js_1 = require("./coin/sc.js");
Object.defineProperty(exports, "sc", { enumerable: true, get: function () { return sc_js_1.sc; } });
var sero_js_1 = require("./coin/sero.js");
Object.defineProperty(exports, "sero", { enumerable: true, get: function () { return sero_js_1.sero; } });
var sol_js_1 = require("./coin/sol.js");
Object.defineProperty(exports, "sol", { enumerable: true, get: function () { return sol_js_1.sol; } });
var srm_js_1 = require("./coin/srm.js");
Object.defineProperty(exports, "srm", { enumerable: true, get: function () { return srm_js_1.srm; } });
var steem_js_1 = require("./coin/steem.js");
Object.defineProperty(exports, "steem", { enumerable: true, get: function () { return steem_js_1.steem; } });
var strat_js_1 = require("./coin/strat.js");
Object.defineProperty(exports, "strat", { enumerable: true, get: function () { return strat_js_1.strat; } });
var strk_js_1 = require("./coin/strk.js");
Object.defineProperty(exports, "strk", { enumerable: true, get: function () { return strk_js_1.strk; } });
var stx_js_1 = require("./coin/stx.js");
Object.defineProperty(exports, "stx", { enumerable: true, get: function () { return stx_js_1.stx; } });
var sys_js_1 = require("./coin/sys.js");
Object.defineProperty(exports, "sys", { enumerable: true, get: function () { return sys_js_1.sys; } });
var tfuel_js_1 = require("./coin/tfuel.js");
Object.defineProperty(exports, "tfuel", { enumerable: true, get: function () { return tfuel_js_1.tfuel; } });
var thetaLegacy_js_1 = require("./coin/thetaLegacy.js");
Object.defineProperty(exports, "thetaLegacy", { enumerable: true, get: function () { return thetaLegacy_js_1.thetaLegacy; } });
var tomoLegacy_js_1 = require("./coin/tomoLegacy.js");
Object.defineProperty(exports, "tomoLegacy", { enumerable: true, get: function () { return tomoLegacy_js_1.tomoLegacy; } });
var trx_js_1 = require("./coin/trx.js");
Object.defineProperty(exports, "trx", { enumerable: true, get: function () { return trx_js_1.trx; } });
var ttLegacy_js_1 = require("./coin/ttLegacy.js");
Object.defineProperty(exports, "ttLegacy", { enumerable: true, get: function () { return ttLegacy_js_1.ttLegacy; } });
var vet_js_1 = require("./coin/vet.js");
Object.defineProperty(exports, "vet", { enumerable: true, get: function () { return vet_js_1.vet; } });
var via_js_1 = require("./coin/via.js");
Object.defineProperty(exports, "via", { enumerable: true, get: function () { return via_js_1.via; } });
var vlx_js_1 = require("./coin/vlx.js");
Object.defineProperty(exports, "vlx", { enumerable: true, get: function () { return vlx_js_1.vlx; } });
var vlxLegacy_js_1 = require("./coin/vlxLegacy.js");
Object.defineProperty(exports, "vlxLegacy", { enumerable: true, get: function () { return vlxLegacy_js_1.vlxLegacy; } });
var vsys_js_1 = require("./coin/vsys.js");
Object.defineProperty(exports, "vsys", { enumerable: true, get: function () { return vsys_js_1.vsys; } });
var wan_js_1 = require("./coin/wan.js");
Object.defineProperty(exports, "wan", { enumerable: true, get: function () { return wan_js_1.wan; } });
var waves_js_1 = require("./coin/waves.js");
Object.defineProperty(exports, "waves", { enumerable: true, get: function () { return waves_js_1.waves; } });
var wicc_js_1 = require("./coin/wicc.js");
Object.defineProperty(exports, "wicc", { enumerable: true, get: function () { return wicc_js_1.wicc; } });
var xch_js_1 = require("./coin/xch.js");
Object.defineProperty(exports, "xch", { enumerable: true, get: function () { return xch_js_1.xch; } });
var xem_js_1 = require("./coin/xem.js");
Object.defineProperty(exports, "xem", { enumerable: true, get: function () { return xem_js_1.xem; } });
var xhv_js_1 = require("./coin/xhv.js");
Object.defineProperty(exports, "xhv", { enumerable: true, get: function () { return xhv_js_1.xhv; } });
var xlm_js_1 = require("./coin/xlm.js");
Object.defineProperty(exports, "xlm", { enumerable: true, get: function () { return xlm_js_1.xlm; } });
var xmr_js_1 = require("./coin/xmr.js");
Object.defineProperty(exports, "xmr", { enumerable: true, get: function () { return xmr_js_1.xmr; } });
var xrp_js_1 = require("./coin/xrp.js");
Object.defineProperty(exports, "xrp", { enumerable: true, get: function () { return xrp_js_1.xrp; } });
var xtz_js_1 = require("./coin/xtz.js");
Object.defineProperty(exports, "xtz", { enumerable: true, get: function () { return xtz_js_1.xtz; } });
var xvg_js_1 = require("./coin/xvg.js");
Object.defineProperty(exports, "xvg", { enumerable: true, get: function () { return xvg_js_1.xvg; } });
var zec_js_1 = require("./coin/zec.js");
Object.defineProperty(exports, "zec", { enumerable: true, get: function () { return zec_js_1.zec; } });
var zen_js_1 = require("./coin/zen.js");
Object.defineProperty(exports, "zen", { enumerable: true, get: function () { return zen_js_1.zen; } });
var zil_js_1 = require("./coin/zil.js");
Object.defineProperty(exports, "zil", { enumerable: true, get: function () { return zil_js_1.zil; } });
