// rocketh/dist/utils/eth.js
function avg(arr) {
    const sum = arr.reduce((a, v) => a + v);
    return sum / BigInt(arr.length);
}
export async function getGasPriceEstimate(provider, options) {
    const defaultOptions = {
        blockCount: 20,
        newestBlock: 'pending',
        rewardPercentiles: [10, 50, 80],
    };
    const optionsResolved = options ? { ...defaultOptions, ...options } : defaultOptions;
    const historicalBlocks = optionsResolved.blockCount;
    const rawFeeHistory = await provider.request({
        method: 'eth_feeHistory',
        params: [`0x${historicalBlocks.toString(16)}`, optionsResolved.newestBlock, optionsResolved.rewardPercentiles],
    });
    let blockNum = Number(rawFeeHistory.oldestBlock);
    const lastBlock = blockNum + rawFeeHistory.reward.length;
    let index = 0;
    const blocksHistory = [];
    while (blockNum < lastBlock) {
        blocksHistory.push({
            number: blockNum,
            baseFeePerGas: BigInt(rawFeeHistory.baseFeePerGas[index]),
            gasUsedRatio: Number(rawFeeHistory.gasUsedRatio[index]),
            priorityFeePerGas: rawFeeHistory.reward[index].map((x) => BigInt(x)),
        });
        blockNum += 1;
        index += 1;
    }
    const percentilePriorityFeeAverages = [];
    for (let i = 0; i < optionsResolved.rewardPercentiles.length; i++) {
        percentilePriorityFeeAverages.push(avg(blocksHistory.map((b) => b.priorityFeePerGas[i])));
    }
    const baseFeePerGas = BigInt(rawFeeHistory.baseFeePerGas[rawFeeHistory.baseFeePerGas.length - 1]);
    const result = [];
    for (let i = 0; i < optionsResolved.rewardPercentiles.length; i++) {
        result.push({
            maxFeePerGas: percentilePriorityFeeAverages[i] + baseFeePerGas,
            maxPriorityFeePerGas: percentilePriorityFeeAverages[i],
        });
    }
    return result;
}
export async function getRoughGasPriceEstimate(provider, options) {
    const defaultOptions = {
        blockCount: 20,
        newestBlock: 'pending',
        rewardPercentiles: [10, 50, 80],
    };
    const optionsResolved = options ? { ...defaultOptions, ...options } : defaultOptions;
    if (optionsResolved.rewardPercentiles.length !== 3) {
        throw new Error(`rough gas estimate require 3 percentile, it defaults to [10,50,80]`);
    }
    const result = await getGasPriceEstimate(provider, optionsResolved);
    return {
        slow: result[0],
        average: result[1],
        fast: result[2],
    };
}
